### DISCUSSION TITLE
Export setting

### DISCUSSION THREAD

**Discussion by blondie63** [Laravel-Backpack/community-forum#585] [Q&A (Help)] [answered]
Hi, i've enabled adding CRUD::enableExportButtons(); and i've two questions:
1. by default export button export only the records on page but this is not so logic because normally i want to list my records choosing a filter but i need to export all records! Is there a setting for this ?
2. when i choose export as PDF i have something like this:
<img width="650" alt="bkp-export-pdf" src="https://github.com/Laravel-Backpack/community-forum/assets/1261851/b597fc03-5c19-4077-ac09-d29006614a10">
but how can i choose the page format (A4, A3, custom) and columns header and content alignment ? for example i'd like to define for each column one alignment for header and different one for value, is it possible ? 

Thanks for help

**Comment by blondie63** [comment]
ps. At now i'm still in BKP v5:
### PHP VERSION:
PHP 8.2.7 (cli) (built: Jun 17 2023 08:16:17) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.2.7, Copyright (c) Zend Technologies
    with Xdebug v3.2.1, Copyright (c) 2002-2023, by Derick Rethans
    with Zend OPcache v8.2.7, Copyright (c), by Zend Technologies

### LARAVEL VERSION:
v10.15.0@c7599dc92e04532824bafbd226c2936ce6a905b8

### BACKPACK PACKAGE VERSIONS:
backpack/crud: 5.6.1
backpack/devtools: 1.3.5
backpack/editable-columns: 2.1.5
backpack/generators: v3.3.17
backpack/pro: 1.6.6
backpack/settings: 3.1.0

and i'm seeing that 'visibleInExport' => false is not working, is this a bug? is this working fine on v6 ?

**Comment by tabacitu** [comment]
> by default export button export only the records on page but this is not so logic because normally i want to list my records choosing a filter but i need to export all records! Is there a setting for this ?

Indeed it's a limitation of the current export. It only exports VISIBLE ROWS. That's because the export functionality is provided by DataTables. You can work around it easily as an admin, by choosing to list "All" rows, not only "10" or "25" or "100".

> when i choose export as PDF i have something like this:

This isn't something that Backpack itself controls, it's something that DataTable does. We'll look into it, but I'm not sure we can do anything about it, unfortunately. I know the PDF export isn't pretty, it's something I don't like either but it hasn't been high on our priority list.

---

We have been working on an alternative Export functionality. An "ExportOperation", that would work in ALL situations, but we've hit a big bottleneck there too. It works ok for 10.000 rows for example, but not for 1 million rows. So it's not a complete solution either. So rather than putting out _yet another_ incomplete solution, we've decided to shelve the issue and package for some time, until we can find a permanent and complete solution (move the processing to a background process and maybe a queue). But that increases the size of the task to months of work, which we couldn't do during the v6 launch, there just wasn't enough time. We'll probably get time to take another look at the ExportOperation and finish it in the next few months. But until then, we'll have to make due with what we have - the DataTables export. I know it isn't perfect, but it has served us well for 7+ years so for most projects it should be "good enough".

Cheers!

**Reply by blondie63** [reply]
ok thanks, what about 'visibleInExport' => false is not working ?

**Reply by tabacitu** [reply]
I'm going to ask @jorgetwgroup to take a look at that, thank you. https://github.com/Laravel-Backpack/community-forum/issues/602

**Reply by jcastroa87** [reply, marked as answer]
Hello @blondie63 

```
'visibleInExport' => false,
```

is working as expected 

I make a test with Monster on demo:
![image](https://github.com/Laravel-Backpack/community-forum/assets/44643608/353f1e75-d749-4a30-bf77-e6aa4b0bdf6b)

With normal params:
XLSX
<img width="1084" alt="image" src="https://github.com/Laravel-Backpack/community-forum/assets/44643608/a5c45d2d-45ef-4590-af58-330773c187ad">

PDF
<img width="937" alt="image" src="https://github.com/Laravel-Backpack/community-forum/assets/44643608/0d09d471-9afc-4a27-83bd-2eb614a166a5">


Now i change this:
```
[
  'name'  => 'checkbox',
  'label' => 'Boolean',
  'type'  => 'boolean',
  // optionally override the Yes/No texts
  'options' => [0 => 'Yes', 1 => 'No'],
  'wrapper' => [
      'element' => 'span',
      'class'   => function ($crud, $column, $entry, $related_key) {
          if ($column['text'] == 'Yes') {
              return 'badge rounded-pill bg-success';
          }

          return 'badge rounded-pill bg-danger';
      },
  ],
  'visibleInExport' => false,
],
```

So "Boolean" column need to be hide:
XLSX
<img width="1026" alt="image" src="https://github.com/Laravel-Backpack/community-forum/assets/44643608/1dbe9a90-424c-4095-8e52-0300ebcfdd11">

PDF
<img width="750" alt="image" src="https://github.com/Laravel-Backpack/community-forum/assets/44643608/eedbba14-4d4f-4915-9b8b-44f772535b47">

Is working correctly.

Let us know how did you try it when fails for you, thanks.

Cheers.
